pkg/storage/unified/resource/table.go

```diff
@@ -0,0 +1,558 @@
+		}
+		columns[i] = *col
+		table.ColumnDefinitions[i] = metav1.TableColumnDefinition{
+			Name:        c.Name,
+			Description: c.Description,
+			Priority:    c.Priority,
+			Type:        col.OpenAPIType,
+			Format:      col.OpenAPIFormat,
+		}
+	}
+
+	var err error
+	table.Rows = make([]metav1.TableRow, len(x.Rows))
+	for i, r := range x.Rows {
+		row := metav1.TableRow{
+			Cells: make([]interface{}, len(r.Cells)),
+		}
+		if len(r.Cells) != columnCount {
+			return table, fmt.Errorf("invalid cells size (have=%d, expect=%d)", len(r.Cells), columnCount)
+		}
+
+		for j, v := range r.Cells {
+			row.Cells[j], err = columns[j].Decode(v)
+			if err != nil {
+				col := columns[j]
+				return table, fmt.Errorf("error decoding (row=%d, column=%d, type=%s) %w", i, j, col.def.Type.String(), err)
+			}
+		}
+
+		// The raw object value
+		if r.Object != nil {
+			row.Object = runtime.RawExtension{
+				Raw: r.Object,
+			}
+		} else if r.Key != nil {
+			obj := &metav1.PartialObjectMetadata{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      r.Key.Name,
+					Namespace: r.Key.Namespace,
+				},
+			}
+			if r.ResourceVersion > 0 {
+				obj.ResourceVersion = strconv.FormatInt(r.ResourceVersion, 10)
+			}
+			row.Object.Object = obj
+			row.Object.Raw, err = json.Marshal(obj)
+			if err != nil {
+				return table, err
+			}
+		}
+		table.Rows[i] = row
+	}
+	return table, err
+}
+
+type TableBuilder struct {
+	ResourceTable
+
+	lookup map[string]*resourceTableColumn
+
+	// Just keep track of it
+	hasDuplicateNames bool
+
+	// When add row gets a colum
+	ignoreUnknownColumns bool
+}
+
+func NewTableBuilder(cols []*ResourceTableColumnDefinition) (*TableBuilder, error) {
+	table := &TableBuilder{
+		ResourceTable: ResourceTable{
+			Columns: cols,
+		},
+
+		lookup: make(map[string]*resourceTableColumn, len(cols)),
+
+		// defaults
+		ignoreUnknownColumns: false,
+	}
+	var err error
+	for i, v := range cols {
+		if table.lookup[v.Name] != nil {
+			table.hasDuplicateNames = true
+			continue
+		}
+		table.lookup[v.Name], err = newResourceTableColumn(v, i)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return table, err
+}
+
+func (x *TableBuilder) AddRow(key *ResourceKey, rv int64, vals map[string]any) error {
+	row := &ResourceTableRow{
+		Key:             key,
+		ResourceVersion: rv,
+		Cells:           make([][]byte, len(x.Columns)),
+	}
+
+	for k, v := range vals {
+		column, ok := x.lookup[k]
+		if !ok {
+			if x.ignoreUnknownColumns {
+				continue
+			}
+			return fmt.Errorf("unknown column")
+		}
+		b, err := column.Encode(v)
+		if err != nil {
+			return err
+		}
+		row.Cells[column.index] = b
+	}
+
+	x.Rows = append(x.Rows, row)
+	return nil
+}
+
+type resourceTableColumn struct {
+	def   *ResourceTableColumnDefinition
+	index int
+
+	// Used for array indexing
+	reader func(iter *jsoniter.Iterator) (any, error)
+	writer func(v any, stream *jsoniter.Stream) error
+
+	OpenAPIType   string
+	OpenAPIFormat string
+}
+
+// nolint:gocyclo
```

Comment:
this is the complicated function -- for each type it picks how to encode/decode efficiently

There is likely a cleaner way to structure this, but it is all internal so maybe good enough to work with for now?